File: frontend/web-nextjs/src/services/categoryService.ts
import { RestResponse } from '@/types/api';

// Types for Category API responses
export interface CategoryItem {
  value: string;
  displayName: string;
  mainCategory: string;
  subCategory: string;
  specificCategory: string;
  description: string;
}

export interface CategoryResponse {
  categories: CategoryItem[];
  mainCategories: string[];
  totalCount: number;
}

export interface StatusItem {
  value: string;
  displayName: string;
  description: string;
  editable: boolean;
  active: boolean;
  finalStatus: boolean;
}

export interface StatusResponse {
  statuses: StatusItem[];
  totalCount: number;
}

export interface TypeItem {
  value: string;
  displayName: string;
  description: string;
  group: string;
}

export interface TypeResponse {
  types: TypeItem[];
  groups: string[];
  totalCount: number;
}

export interface CategoryStats {
  totalCategories: number;
  totalMainCategories: number;
  totalStatuses: number;
  totalTypes: number;
  totalGroups: number;
}

/**
 * Service for managing categories, statuses, and types
 */
class CategoryService {
  private baseUrl = '/api/v1/contract-management-service/categories';

  /**
   * Get all categories
   */
  async getAllCategories(): Promise<RestResponse<CategoryResponse>> {
    const response = await fetch(`${this.baseUrl}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch categories: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get categories by main category
   */
  async getCategoriesByMainCategory(mainCategory: string): Promise<RestResponse<CategoryResponse>> {
    const response = await fetch(`${this.baseUrl}/${encodeURIComponent(mainCategory)}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch categories for main category ${mainCategory}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get categories by sub category
   */
  async getCategoriesBySubCategory(
    mainCategory: string, 
    subCategory: string
  ): Promise<RestResponse<CategoryResponse>> {
    const response = await fetch(
      `${this.baseUrl}/${encodeURIComponent(mainCategory)}/${encodeURIComponent(subCategory)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch categories for sub category ${subCategory}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Search categories
   */
  async searchCategories(query: string): Promise<RestResponse<CategoryResponse>> {
    const response = await fetch(`${this.baseUrl}/search?query=${encodeURIComponent(query)}`);
    if (!response.ok) {
      throw new Error(`Failed to search categories: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get all statuses
   */
  async getAllStatuses(): Promise<RestResponse<StatusResponse>> {
    const response = await fetch(`${this.baseUrl}/statuses`);
    if (!response.ok) {
      throw new Error(`Failed to fetch statuses: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get possible next statuses from current status
   */
  async getPossibleNextStatuses(currentStatus: string): Promise<RestResponse<StatusResponse>> {
    const response = await fetch(`${this.baseUrl}/statuses/${encodeURIComponent(currentStatus)}/next`);
    if (!response.ok) {
      throw new Error(`Failed to fetch next statuses for ${currentStatus}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get all types
   */
  async getAllTypes(): Promise<RestResponse<TypeResponse>> {
    const response = await fetch(`${this.baseUrl}/types`);
    if (!response.ok) {
      throw new Error(`Failed to fetch types: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get types by group
   */
  async getTypesByGroup(group: string): Promise<RestResponse<TypeResponse>> {
    const response = await fetch(`${this.baseUrl}/types/group/${encodeURIComponent(group)}`);
    if (!response.ok) {
      throw new Error(`Failed to fetch types for group ${group}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Search types
   */
  async searchTypes(query: string): Promise<RestResponse<TypeResponse>> {
    const response = await fetch(`${this.baseUrl}/types/search?query=${encodeURIComponent(query)}`);
    if (!response.ok) {
      throw new Error(`Failed to search types: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Validate category
   */
  async validateCategory(category: string): Promise<RestResponse<boolean>> {
    const response = await fetch(`${this.baseUrl}/validate/${encodeURIComponent(category)}`);
    if (!response.ok) {
      throw new Error(`Failed to validate category ${category}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Validate status
   */
  async validateStatus(status: string): Promise<RestResponse<boolean>> {
    const response = await fetch(`${this.baseUrl}/statuses/validate/${encodeURIComponent(status)}`);
    if (!response.ok) {
      throw new Error(`Failed to validate status ${status}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Validate type
   */
  async validateType(type: string): Promise<RestResponse<boolean>> {
    const response = await fetch(`${this.baseUrl}/types/validate/${encodeURIComponent(type)}`);
    if (!response.ok) {
      throw new Error(`Failed to validate type ${type}: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Validate status transition
   */
  async validateStatusTransition(
    currentStatus: string, 
    newStatus: string
  ): Promise<RestResponse<boolean>> {
    const response = await fetch(
      `${this.baseUrl}/statuses/validate-transition?currentStatus=${encodeURIComponent(currentStatus)}&newStatus=${encodeURIComponent(newStatus)}`
    );
    if (!response.ok) {
      throw new Error(`Failed to validate status transition: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get category statistics
   */
  async getCategoryStats(): Promise<RestResponse<CategoryStats>> {
    const response = await fetch(`${this.baseUrl}/stats`);
    if (!response.ok) {
      throw new Error(`Failed to fetch category stats: ${response.statusText}`);
    }
    return response.json();
  }

  /**
   * Get hierarchical categories (main -> sub -> specific)
   */
  async getHierarchicalCategories(): Promise<{
    [mainCategory: string]: {
      [subCategory: string]: CategoryItem[];
    };
  }> {
    const response = await this.getAllCategories();
    const hierarchical: { [mainCategory: string]: { [subCategory: string]: CategoryItem[] } } = {};

    response.data.categories.forEach(category => {
      if (!hierarchical[category.mainCategory]) {
        hierarchical[category.mainCategory] = {};
      }
      if (!hierarchical[category.mainCategory][category.subCategory]) {
        hierarchical[category.mainCategory][category.subCategory] = [];
      }
      hierarchical[category.mainCategory][category.subCategory].push(category);
    });

    return hierarchical;
  }

  /**
   * Get categories grouped by main category
   */
  async getCategoriesGroupedByMain(): Promise<{ [mainCategory: string]: CategoryItem[] }> {
    const response = await this.getAllCategories();
    const grouped: { [mainCategory: string]: CategoryItem[] } = {};

    response.data.categories.forEach(category => {
      if (!grouped[category.mainCategory]) {
        grouped[category.mainCategory] = [];
      }
      grouped[category.mainCategory].push(category);
    });

    return grouped;
  }

  /**
   * Get types grouped by group
   */
  async getTypesGroupedByGroup(): Promise<{ [group: string]: TypeItem[] }> {
    const response = await this.getAllTypes();
    const grouped: { [group: string]: TypeItem[] } = {};

    response.data.types.forEach(type => {
      if (!grouped[type.group]) {
        grouped[type.group] = [];
      }
      grouped[type.group].push(type);
    });

    return grouped;
  }

  /**
   * Search across all categories, statuses, and types
   */
  async searchAll(query: string): Promise<{
    categories: CategoryItem[];
    statuses: StatusItem[];
    types: TypeItem[];
  }> {
    const [categoriesResponse, statusesResponse, typesResponse] = await Promise.all([
      this.searchCategories(query),
      this.getAllStatuses(),
      this.searchTypes(query)
    ]);

    // Filter statuses by query
    const filteredStatuses = statusesResponse.data.statuses.filter(status =>
      status.value.toLowerCase().includes(query.toLowerCase()) ||
      status.displayName.toLowerCase().includes(query.toLowerCase()) ||
      status.description.toLowerCase().includes(query.toLowerCase())
    );

    return {
      categories: categoriesResponse.data.categories,
      statuses: filteredStatuses,
      types: typesResponse.data.types
    };
  }
}

// Export singleton instance
export const categoryService = new CategoryService();
export default categoryService;

File: backend/document-management-service/src/main/java/com/devgo2003/docgo/document_service/service/CategoryService.java
package com.devgo2003.docgo.document_service.service;

import com.devgo2003.docgo.document_service.dto.CategoryResponse;
import com.devgo2003.docgo.document_service.enums.ContractCategory;
import com.devgo2003.docgo.document_service.enums.ContractStatus;
import com.devgo2003.docgo.document_service.enums.ContractType;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service cho quản lý categories, statuses và types
 */
@Service
public class CategoryService {
    
    /**
     * Lấy tất cả categories
     */
    @Cacheable(value = "categories", key = "'all'")
    public CategoryResponse getAllCategories() {
        List<CategoryResponse.CategoryItem> categories = Arrays.stream(ContractCategory.values())
                .map(CategoryResponse.CategoryItem::fromContractCategory)
                .collect(Collectors.toList());
        
        List<String> mainCategories = Arrays.stream(ContractCategory.getMainCategories())
                .collect(Collectors.toList());
        
        return new CategoryResponse(categories, mainCategories, categories.size());
    }
    
    /**
     * Lấy categories theo main category
     */
    @Cacheable(value = "categories", key = "#mainCategory")
    public CategoryResponse getCategoriesByMainCategory(String mainCategory) {
        ContractCategory[] categories = ContractCategory.getByMainCategory(mainCategory);
        
        List<CategoryResponse.CategoryItem> categoryItems = Arrays.stream(categories)
                .map(CategoryResponse.CategoryItem::fromContractCategory)
                .collect(Collectors.toList());
        
        List<String> subCategories = Arrays.stream(ContractCategory.getSubCategories(mainCategory))
                .collect(Collectors.toList());
        
        return new CategoryResponse(categoryItems, subCategories, categoryItems.size());
    }
    
    /**
     * Lấy categories theo sub category
     */
    @Cacheable(value = "categories", key = "#mainCategory + '_' + #subCategory")
    public CategoryResponse getCategoriesBySubCategory(String mainCategory, String subCategory) {
        ContractCategory[] categories = ContractCategory.getBySubCategory(mainCategory, subCategory);
        
        List<CategoryResponse.CategoryItem> categoryItems = Arrays.stream(categories)
                .map(CategoryResponse.CategoryItem::fromContractCategory)
                .collect(Collectors.toList());
        
        return new CategoryResponse(categoryItems, null, categoryItems.size());
    }
    
    /**
     * Tìm kiếm categories
     */
    public CategoryResponse searchCategories(String keyword) {
        ContractCategory[] categories = ContractCategory.search(keyword);
        
        List<CategoryResponse.CategoryItem> categoryItems = Arrays.stream(categories)
                .map(CategoryResponse.CategoryItem::fromContractCategory)
                .collect(Collectors.toList());
        
        return new CategoryResponse(categoryItems, null, categoryItems.size());
    }
    
    /**
     * Lấy tất cả statuses
     */
    @Cacheable(value = "statuses", key = "'all'")
    public CategoryResponse.StatusResponse getAllStatuses() {
        List<CategoryResponse.StatusResponse.StatusItem> statuses = Arrays.stream(ContractStatus.values())
                .map(CategoryResponse.StatusResponse.StatusItem::fromContractStatus)
                .collect(Collectors.toList());
        
        return new CategoryResponse.StatusResponse(statuses, statuses.size());
    }
    
    /**
     * Lấy statuses có thể chuyển từ status hiện tại
     */
    public CategoryResponse.StatusResponse getPossibleNextStatuses(String currentStatus) {
        try {
            ContractStatus status = ContractStatus.fromValue(currentStatus);
            ContractStatus[] possibleStatuses = status.getPossibleNextStatuses();
            
            List<CategoryResponse.StatusResponse.StatusItem> statusItems = Arrays.stream(possibleStatuses)
                    .map(CategoryResponse.StatusResponse.StatusItem::fromContractStatus)
                    .collect(Collectors.toList());
            
            return new CategoryResponse.StatusResponse(statusItems, statusItems.size());
        } catch (IllegalArgumentException e) {
            return new CategoryResponse.StatusResponse(List.of(), 0);
        }
    }
    
    /**
     * Lấy tất cả types
     */
    @Cacheable(value = "types", key = "'all'")
    public CategoryResponse.TypeResponse getAllTypes() {
        List<CategoryResponse.TypeResponse.TypeItem> types = Arrays.stream(ContractType.values())
                .map(CategoryResponse.TypeResponse.TypeItem::fromContractType)
                .collect(Collectors.toList());
        
        List<String> groups = Arrays.stream(ContractType.getGroups())
                .collect(Collectors.toList());
        
        return new CategoryResponse.TypeResponse(types, groups, types.size());
    }
    
    /**
     * Lấy types theo group
     */
    @Cacheable(value = "types", key = "#group")
    public CategoryResponse.TypeResponse getTypesByGroup(String group) {
        ContractType[] types = ContractType.getByGroup(group);
        
        List<CategoryResponse.TypeResponse.TypeItem> typeItems = Arrays.stream(types)
                .map(CategoryResponse.TypeResponse.TypeItem::fromContractType)
                .collect(Collectors.toList());
        
        return new CategoryResponse.TypeResponse(typeItems, null, typeItems.size());
    }
    
    /**
     * Tìm kiếm types
     */
    public CategoryResponse.TypeResponse searchTypes(String keyword) {
        ContractType[] types = ContractType.search(keyword);
        
        List<CategoryResponse.TypeResponse.TypeItem> typeItems = Arrays.stream(types)
                .map(CategoryResponse.TypeResponse.TypeItem::fromContractType)
                .collect(Collectors.toList());
        
        return new CategoryResponse.TypeResponse(typeItems, null, typeItems.size());
    }
    
    /**
     * Validate category có tồn tại không
     */
    public boolean isValidCategory(String category) {
        try {
            ContractCategory.fromValue(category);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
    
    /**
     * Validate status có tồn tại không
     */
    public boolean isValidStatus(String status) {
        try {
            ContractStatus.fromValue(status);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
    
    /**
     * Validate type có tồn tại không
     */
    public boolean isValidType(String type) {
        try {
            ContractType.fromValue(type);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
    
    /**
     * Validate transition từ status hiện tại sang status mới
     */
    public boolean isValidStatusTransition(String currentStatus, String newStatus) {
        try {
            ContractStatus current = ContractStatus.fromValue(currentStatus);
            ContractStatus target = ContractStatus.fromValue(newStatus);
            return current.canTransitionTo(target);
        } catch (IllegalArgumentException e) {
            return false;
        }
    }
    
    /**
     * Lấy thống kê categories
     */
    public CategoryStats getCategoryStats() {
        return new CategoryStats(
            ContractCategory.values().length,
            ContractCategory.getMainCategories().length,
            ContractStatus.values().length,
            ContractType.values().length,
            ContractType.getGroups().length
        );
    }
    
    /**
     * Inner class cho Category Stats
     */
    public static class CategoryStats {
        private final int totalCategories;
        private final int totalMainCategories;
        private final int totalStatuses;
        private final int totalTypes;
        private final int totalGroups;
        
        public CategoryStats(int totalCategories, int totalMainCategories, 
                           int totalStatuses, int totalTypes, int totalGroups) {
            this.totalCategories = totalCategories;
            this.totalMainCategories = totalMainCategories;
            this.totalStatuses = totalStatuses;
            this.totalTypes = totalTypes;
            this.totalGroups = totalGroups;
        }
        
        // Getters
        public int getTotalCategories() {
            return totalCategories;
        }
        
        public int getTotalMainCategories() {
            return totalMainCategories;
        }
        
        public int getTotalStatuses() {
            return totalStatuses;
        }
        
        public int getTotalTypes() {
            return totalTypes;
        }
        
        public int getTotalGroups() {
            return totalGroups;
        }
    }
}

